### ISSUE TITLE
Upstream retry is performed on an outlier host

### ISSUE THREAD

**Issue by ohadvano** [envoyproxy/envoy#31453]
I have a setup of downstream ``tcp_proxy`` and upstream ``STRICT_DNS`` cluster:
```
static_resources:
  listeners:
  - name: main
    stat_prefix: main
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 300
    filter_chains:
    - name: main
      filters:
      - name: tcp_proxy
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
          stat_prefix: tcp
          cluster: example_cluster
          max_connect_attempts: 20
          tunneling_config:
            hostname: host.com
            use_post: true
            post_path: /

  clusters:
  - name: example_cluster
    type: STRICT_DNS
    outlier_detection:
      consecutive_5xx: 1
      max_ejection_percent: 100
      enforcing_consecutive_5xx: 100
    load_assignment:
      cluster_name: example_cluster
      endpoints:
      - priority: 0
        lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: fakehost1.com
                port_value: 8500
      - priority: 1
        lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: fakehost2.com
                port_value: 8500
      - priority: 2
        lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: realhost.com
                port_value: 8500
```

The ``fakehost1.com`` and ``fakehost2.com`` hosts are real hosts with valid DNS name, but they don't listen on port 8500, so it's expected the upstream connection to fail. ``realhost.com`` is a valid DNS name and also listening on port 8500, so it's expected that connecting to it will work.

With the config above, I'm expecting to see upstream attempts to ``[fakehost1.com, fakehost2.com, realhost.com]``, but in reality I see ``[fakehost1.com, fakehost1.com, fakehost2.com, fakehost2.com, repeating fakehost2.com for as many retries I configure]``.

There are two problems:
- ``fakehost1.com`` is picked at the second attempt, although first attempt failed and it should be ejected sooner
- ``fakehost2.com`` is retried forever, although attempts to it are failing, and logs show that it is ejected by outlier detector

Logs show that the upstream hosts are declared outlier, but still they are being picked:

```
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"0\"] Creating connection to cluster example_cluster]
2023-12-19 18:10:10 [TID: 32][msg: queueing stream due to no available connections (ready=0 busy=0 connecting=0)]
2023-12-19 18:10:10 [TID: 32][msg: trying to create new connection]
2023-12-19 18:10:10 [TID: 32][msg: ConnPoolImplBase 0x2618ffc6ab40, ready_clients_.size(): 0, busy_clients_.size(): 0, connecting_clients_.size(): 0, connecting_stream_capacity_: 0, num_active_streams_: 0, pending_streams_.size(): 1 per upstream preconnect ratio: 1]
2023-12-19 18:10:10 [TID: 32][msg: creating a new connection (connecting=0)]
2023-12-19 18:10:10 [TID: 32][msg: Codec does not have Metadata frame support.]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"1\"] updating connection-level initial window size to 1048576]
2023-12-19 18:10:10 [TID: 32][msg: [C1] current connecting state: true]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"1\"] connecting]
2023-12-19 18:10:10 [TID: 32][msg: [C1] connecting to 10.5.0.2:8500]
2023-12-19 18:10:10 [TID: 32][msg: [C1] connection in progress]
2023-12-19 18:10:10 [TID: 32][msg: not creating a new connection, shouldCreateNewConnection returned false.]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"1\"] socket event: 3]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"1\"] write ready]
2023-12-19 18:10:10 [TID: 32][msg: [C1] delayed connect error: 111]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"1\"] closing socket: 0]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"1\"] raising connection event 0]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"1\"] disconnect. resetting 0 pending requests]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"1\"] client disconnected, failure reason: delayed connect error: 111]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"0\"] Creating connection to cluster example_cluster]
2023-12-19 18:10:10 [TID: 32][msg: queueing stream due to no available connections (ready=0 busy=0 connecting=1)]
2023-12-19 18:10:10 [TID: 32][msg: trying to create new connection]
2023-12-19 18:10:10 [TID: 32][msg: ConnPoolImplBase 0x2618ffc6ab40, ready_clients_.size(): 0, busy_clients_.size(): 0, connecting_clients_.size(): 1, connecting_stream_capacity_: 0, num_active_streams_: 0, pending_streams_.size(): 1 per upstream preconnect ratio: 1]
2023-12-19 18:10:10 [TID: 32][msg: creating a new connection (connecting=1)]
2023-12-19 18:10:10 [TID: 25][msg: membership update for TLS cluster example_cluster added 0 removed 0]
2023-12-19 18:10:10 [TID: 25][msg: recalculated priority state: priority level 0, healthy weight 0, total weight 1, overprovision factor 140, healthy result 0, degraded result 0]
2023-12-19 18:10:10 [TID: 33][msg: membership update for TLS cluster example_cluster added 0 removed 0]
2023-12-19 18:10:10 [TID: 33][msg: recalculated priority state: priority level 0, healthy weight 0, total weight 1, overprovision factor 140, healthy result 0, degraded result 0]
2023-12-19 18:10:10 [TID: 25][msg: membership update for TLS cluster example_cluster added 0 removed 0]
2023-12-19 18:10:10 [TID: 25][msg: recalculated priority state: priority level 1, healthy weight 1, total weight 1, overprovision factor 140, healthy result 100, degraded result 0]
2023-12-19 18:10:10 [TID: 33][msg: membership update for TLS cluster example_cluster added 0 removed 0]
2023-12-19 18:10:10 [TID: 33][msg: recalculated priority state: priority level 1, healthy weight 1, total weight 1, overprovision factor 140, healthy result 100, degraded result 0]
2023-12-19 18:10:10 [TID: 32][msg: Codec does not have Metadata frame support.]
2023-12-19 18:10:10 [TID: 25][msg: membership update for TLS cluster example_cluster added 0 removed 0]
2023-12-19 18:10:10 [TID: 25][msg: recalculated priority state: priority level 2, healthy weight 1, total weight 1, overprovision factor 140, healthy result 100, degraded result 0]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"2\"] updating connection-level initial window size to 1048576]
2023-12-19 18:10:10 [TID: 33][msg: membership update for TLS cluster example_cluster added 0 removed 0]
2023-12-19 18:10:10 [TID: 25][msg: host 10.5.0.2:8500 in cluster example_cluster was ejected by the outlier detector]
2023-12-19 18:10:10 [TID: 32][msg: [C2] current connecting state: true]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"2\"] connecting]
2023-12-19 18:10:10 [TID: 33][msg: recalculated priority state: priority level 2, healthy weight 1, total weight 1, overprovision factor 140, healthy result 100, degraded result 0]
2023-12-19 18:10:10 [TID: 32][msg: [C2] connecting to 10.5.0.2:8500]
2023-12-19 18:10:10 [TID: 32][msg: [C2] connection in progress]
2023-12-19 18:10:10 [TID: 32][msg: not creating a new connection, shouldCreateNewConnection returned false.]
2023-12-19 18:10:10 [TID: 32][msg: not creating a new connection, shouldCreateNewConnection returned false.]
2023-12-19 18:10:10 [TID: 32][msg: item added to deferred deletion list (size=1)]
2023-12-19 18:10:10 [TID: 32][msg: not creating a new connection, shouldCreateNewConnection returned false.]
2023-12-19 18:10:10 [TID: 32][msg: clearing deferred deletion list (size=1)]
2023-12-19 18:10:10 [TID: 32][msg: membership update for TLS cluster example_cluster added 0 removed 0]
2023-12-19 18:10:10 [TID: 32][msg: recalculated priority state: priority level 0, healthy weight 0, total weight 1, overprovision factor 140, healthy result 0, degraded result 0]
2023-12-19 18:10:10 [TID: 32][msg: membership update for TLS cluster example_cluster added 0 removed 0]
2023-12-19 18:10:10 [TID: 32][msg: recalculated priority state: priority level 1, healthy weight 1, total weight 1, overprovision factor 140, healthy result 100, degraded result 0]
2023-12-19 18:10:10 [TID: 32][msg: membership update for TLS cluster example_cluster added 0 removed 0]
2023-12-19 18:10:10 [TID: 32][msg: recalculated priority state: priority level 2, healthy weight 1, total weight 1, overprovision factor 140, healthy result 100, degraded result 0]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"2\"] socket event: 3]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"2\"] write ready]
2023-12-19 18:10:10 [TID: 32][msg: [C2] delayed connect error: 111]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"2\"] closing socket: 0]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"2\"] raising connection event 0]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"2\"] disconnect. resetting 0 pending requests]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"2\"] client disconnected, failure reason: delayed connect error: 111]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"0\"] Creating connection to cluster example_cluster]
2023-12-19 18:10:10 [TID: 32][msg: queueing stream due to no available connections (ready=0 busy=0 connecting=0)]
2023-12-19 18:10:10 [TID: 32][msg: trying to create new connection]
2023-12-19 18:10:10 [TID: 32][msg: ConnPoolImplBase 0x2618ffc6b860, ready_clients_.size(): 0, busy_clients_.size(): 0, connecting_clients_.size(): 0, connecting_stream_capacity_: 0, num_active_streams_: 0, pending_streams_.size(): 1 per upstream preconnect ratio: 1]
2023-12-19 18:10:10 [TID: 32][msg: creating a new connection (connecting=0)]
2023-12-19 18:10:10 [TID: 32][msg: Codec does not have Metadata frame support.]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"3\"] updating connection-level initial window size to 1048576]
2023-12-19 18:10:10 [TID: 32][msg: [C3] current connecting state: true]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"3\"] connecting]
2023-12-19 18:10:10 [TID: 32][msg: [C3] connecting to 10.5.0.3:8500]
2023-12-19 18:10:10 [TID: 32][msg: [C3] connection in progress]
2023-12-19 18:10:10 [TID: 32][msg: not creating a new connection, shouldCreateNewConnection returned false.]
2023-12-19 18:10:10 [TID: 32][msg: not creating a new connection, shouldCreateNewConnection returned false.]
2023-12-19 18:10:10 [TID: 32][msg: item added to deferred deletion list (size=1)]
2023-12-19 18:10:10 [TID: 32][msg: invoking idle callbacks - is_draining_for_deletion_=false]
2023-12-19 18:10:10 [TID: 32][msg: Erasing idle pool for host fakehost1.com]
2023-12-19 18:10:10 [TID: 32][msg: item added to deferred deletion list (size=2)]
2023-12-19 18:10:10 [TID: 32][msg: Pool container empty for host fakehost1.com, erasing host entry]
2023-12-19 18:10:10 [TID: 32][msg: clearing deferred deletion list (size=2)]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"3\"] socket event: 3]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"3\"] write ready]
2023-12-19 18:10:10 [TID: 32][msg: [C3] delayed connect error: 111]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"3\"] closing socket: 0]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"3\"] raising connection event 0]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"3\"] disconnect. resetting 0 pending requests]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"3\"] client disconnected, failure reason: delayed connect error: 111]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"0\"] Creating connection to cluster example_cluster]
2023-12-19 18:10:10 [TID: 32][msg: queueing stream due to no available connections (ready=0 busy=0 connecting=1)]
2023-12-19 18:10:10 [TID: 32][msg: trying to create new connection]
2023-12-19 18:10:10 [TID: 32][msg: ConnPoolImplBase 0x2618ffc6b860, ready_clients_.size(): 0, busy_clients_.size(): 0, connecting_clients_.size(): 1, connecting_stream_capacity_: 0, num_active_streams_: 0, pending_streams_.size(): 1 per upstream preconnect ratio: 1]
2023-12-19 18:10:10 [TID: 25][msg: host 10.5.0.3:8500 in cluster example_cluster was ejected by the outlier detector]
2023-12-19 18:10:10 [TID: 32][msg: creating a new connection (connecting=1)]
2023-12-19 18:10:10 [TID: 32][msg: Codec does not have Metadata frame support.]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"4\"] updating connection-level initial window size to 1048576]
2023-12-19 18:10:10 [TID: 32][msg: [C4] current connecting state: true]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"4\"] connecting]
2023-12-19 18:10:10 [TID: 32][msg: [C4] connecting to 10.5.0.3:8500]
2023-12-19 18:10:10 [TID: 32][msg: [C4] connection in progress]
2023-12-19 18:10:10 [TID: 32][msg: not creating a new connection, shouldCreateNewConnection returned false.]
2023-12-19 18:10:10 [TID: 32][msg: not creating a new connection, shouldCreateNewConnection returned false.]
2023-12-19 18:10:10 [TID: 32][msg: item added to deferred deletion list (size=1)]
2023-12-19 18:10:10 [TID: 32][msg: not creating a new connection, shouldCreateNewConnection returned false.]
2023-12-19 18:10:10 [TID: 32][msg: clearing deferred deletion list (size=1)]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"4\"] socket event: 3]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"4\"] write ready]
2023-12-19 18:10:10 [TID: 32][msg: [C4] delayed connect error: 111]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"4\"] closing socket: 0]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"4\"] raising connection event 0]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"4\"] disconnect. resetting 0 pending requests]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"4\"] client disconnected, failure reason: delayed connect error: 111]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"0\"] Creating connection to cluster example_cluster]
2023-12-19 18:10:10 [TID: 32][msg: queueing stream due to no available connections (ready=0 busy=0 connecting=1)]
2023-12-19 18:10:10 [TID: 32][msg: trying to create new connection]
2023-12-19 18:10:10 [TID: 32][msg: ConnPoolImplBase 0x2618ffc6b860, ready_clients_.size(): 0, busy_clients_.size(): 0, connecting_clients_.size(): 1, connecting_stream_capacity_: 0, num_active_streams_: 0, pending_streams_.size(): 1 per upstream preconnect ratio: 1]
2023-12-19 18:10:10 [TID: 32][msg: creating a new connection (connecting=1)]
2023-12-19 18:10:10 [TID: 32][msg: Codec does not have Metadata frame support.]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"5\"] updating connection-level initial window size to 1048576]
2023-12-19 18:10:10 [TID: 32][msg: [C5] current connecting state: true]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"5\"] connecting]
2023-12-19 18:10:10 [TID: 32][msg: [C5] connecting to 10.5.0.3:8500]
2023-12-19 18:10:10 [TID: 32][msg: [C5] connection in progress]
2023-12-19 18:10:10 [TID: 32][msg: not creating a new connection, shouldCreateNewConnection returned false.]
2023-12-19 18:10:10 [TID: 32][msg: not creating a new connection, shouldCreateNewConnection returned false.]
2023-12-19 18:10:10 [TID: 32][msg: item added to deferred deletion list (size=1)]
2023-12-19 18:10:10 [TID: 32][msg: not creating a new connection, shouldCreateNewConnection returned false.]
2023-12-19 18:10:10 [TID: 32][msg: clearing deferred deletion list (size=1)]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"5\"] socket event: 3]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"5\"] write ready]
2023-12-19 18:10:10 [TID: 32][msg: [C5] delayed connect error: 111]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"5\"] closing socket: 0]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"5\"] raising connection event 0]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"5\"] disconnect. resetting 0 pending requests]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"5\"] client disconnected, failure reason: delayed connect error: 111]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"0\"] Creating connection to cluster example_cluster]
2023-12-19 18:10:10 [TID: 32][msg: queueing stream due to no available connections (ready=0 busy=0 connecting=1)]
2023-12-19 18:10:10 [TID: 32][msg: trying to create new connection]
2023-12-19 18:10:10 [TID: 32][msg: ConnPoolImplBase 0x2618ffc6b860, ready_clients_.size(): 0, busy_clients_.size(): 0, connecting_clients_.size(): 1, connecting_stream_capacity_: 0, num_active_streams_: 0, pending_streams_.size(): 1 per upstream preconnect ratio: 1]
2023-12-19 18:10:10 [TID: 32][msg: creating a new connection (connecting=1)]
2023-12-19 18:10:10 [TID: 32][msg: Codec does not have Metadata frame support.]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"6\"] updating connection-level initial window size to 1048576]
2023-12-19 18:10:10 [TID: 32][msg: [C6] current connecting state: true]
2023-12-19 18:10:10 [TID: 32][msg: [Tags: \"ConnectionId\":\"6\"] connecting]
2023-12-19 18:10:10 [TID: 32][msg: [C6] connecting to 10.5.0.3:8500]
```

Host to IP mapping:
``fakehost1.com`` -> 10.5.0.2
``fakehost2.com`` -> 10.5.0.3
``realhost.com``    -> 10.5.0.4

cc @suniltheta

**Comment by ohadvano** [comment]
After trying to debug this, I believe what happens is that the worker thread puts the outlier result and posts to the main thread, to sync the result back to all worker threads. So while the worker thread tries to retry the upstream request, it doesn't always know that the host is unhealthy because the main thread didn't post the update back to the worker threads.

I hope I got this right, but if that's the case - is it not possible to, in parallel, update the health state for the host in the same worker thread the request was originated right after the request has failed (and the main thread update would basically do nothing for that specific thread)?

**Comment by wbpcode** [comment]
nope, for now. It's similar with the active health checker, it need a little time to sync the status. Locking is expensive for this requirement
